tools/pycafe/create_pycafe_links.py

```diff
@@ -0,0 +1,131 @@
+"""Generate PyCafe links for the example dashboards and post them as a comment on the pull request and as status."""
+
+import base64
+import datetime
+import gzip
+import json
+import os
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from typing import Optional
+from urllib.parse import quote, urlencode
+
+from github import Auth, Github
+
+GITHUB_TOKEN = str(os.getenv("GITHUB_TOKEN"))
+REPO_NAME = str(os.getenv("GITHUB_REPOSITORY"))
+PR_NUMBER = int(os.getenv("PR_NUMBER"))
+RUN_ID = str(os.getenv("RUN_ID"))
+PACKAGE_VERSION = subprocess.check_output(["hatch", "version"]).decode("utf-8").strip()
+PYCAFE_URL = "https://py.cafe"
+
+# Access
+auth = Auth.Token(GITHUB_TOKEN)
+g = Github(auth=auth)
+
+# Get PR and commits
+repo = g.get_repo(REPO_NAME)
+pr = repo.get_pull(PR_NUMBER)
+commit_sha = pr.head.sha
+commit = repo.get_commit(commit_sha)
+
+
+def generate_link(directory: str, extra_requirements: Optional[list[str]] = None):
+    """Generate a PyCafe link for the example dashboards."""
+    base_url = f"https://raw.githubusercontent.com/mckinsey/vizro/{commit_sha}/vizro-core/{directory}"
+
+    # Requirements
+    extra_requirements_concat = "\n".join(extra_requirements) if extra_requirements else ""
+    requirements = (
+        f"""{PYCAFE_URL}/gh/artifact/mckinsey/vizro/actions/runs/{RUN_ID}/pip/vizro-{PACKAGE_VERSION}-py3-none-any.whl\n"""
+        + extra_requirements_concat
+    )
+
+    # App file
+    app_file_path = os.path.join(directory, "app.py")
+    app_content = Path(app_file_path).read_text()
+    app_content_split = app_content.split('if __name__ == "__main__":')
+    if len(app_content_split) > 1:
+        app_content = app_content_split[0] + textwrap.dedent(app_content_split[1])
+
+    # JSON object
+    json_object = {
+        "code": str(app_content),
+        "requirements": requirements,
+        "files": [],
+    }
+    for root, _, files in os.walk("./" + directory):
+        for file in files:
+            if "app.py" in file:
+                continue
+            file_path = os.path.join(root, file)
+            relative_path = os.path.relpath(file_path, directory)
+            file_url = f"{base_url}{relative_path.replace(os.sep, '/')}"
+            json_object["files"].append({"name": relative_path, "url": file_url})
+
+    json_text = json.dumps(json_object)
+    compressed_json_text = gzip.compress(json_text.encode("utf8"))
+    base64_text = base64.b64encode(compressed_json_text).decode("utf8")
+    query = urlencode({"c": base64_text}, quote_via=quote)
```

Comment:
A suggestion/feedback for @maartenbreddels. It might be nice if there were some way of generating this with `import pycafe; pycafe.generate_snippet()` or whatever. Unless this compression scheme and URL format will never ever change.